IXFR out

wcawijngaards opened the pull request:
IXFR support to provide IXFR going out, towards requestors. The implementation can store downloaded IXFRs, with `store-ixfr: yes` and create IXFRs from zonefiles with `create-ixfr: yes`.

The `ixfr-number:5` and `ixfr-size: 1048576` settings specify the maximum amount to store for IXFR data.

IXFRs are stored next to the zonefile in '.ixfr' files.

NSD already supports downloading IXFRs and using them, this change also makes it possible to provide IXFRs to requestors.

wcawijngaards committed:
- ixfr-out, data structure.

wcawijngaards committed:
- ixfr-out, store total size and struct collation.

wcawijngaards committed:
- ixfr-out, ixfr store routines that store IXFR during difffile processing.

wcawijngaards committed:
- ixfr-out, ixfr serial query parse routine.

wcawijngaards committed:
- ixfr-out, ixfr check if end serial is current served zone serial.

wcawijngaards committed:
- ixfr-out, ixfr reply

wcawijngaards committed:
- ixfr-out, ixfr tsig sign every packet.

wcawijngaards committed:
- ixfr-out, ixfr add records in sequence.

wcawijngaards committed:
- ixfr-out, ixfr unit test.

wcawijngaards committed:
- ixfr-out, store-ixfr: yesno option that can be set per-zone and in patterns.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, use store-ixfr option in ixfr-out unit test.

wcawijngaards committed:
- ixfr-out, documentation for the store-ixfr option.

wcawijngaards committed:
- ixfr-out, store IXFRs in tree.

wcawijngaards committed:
- ixfr-out, ixfr make space routine to make space for multiple versions.

wcawijngaards committed:
- ixfr-out, ixfr options for ixfr-number and ixfr-size.

wcawijngaards committed:
- ixfr-out, ixfr response for current version implemented as per RFC 1995.

wcawijngaards committed:
- ixfr-out, fix compression domain in error case.

wcawijngaards committed:
- ixfr-out, reply with collated IXFR response from several versions.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, test for IXFR with several versions, and current, future indication

wcawijngaards committed:
- ixfr-out, UDP response for IXFR, RFC1995 compliant, in one packet if it fits
  and truncated if large.

wcawijngaards committed:
- ixfr-out, ixfr_data_read and ixfr_data_readnewsoa routines.

wcawijngaards committed:
- ixfr-out, ixfr_read_from_file.

wcawijngaards committed:
- ixfr-out, fix so that ixfr read from file works.

wcawijngaards committed:
- ixfr-out, comments added.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, keep track of file_num.

wcawijngaards committed:
- ixfr-out, implement ixfr_write_to_file.

wcawijngaards committed:
- ixfr-out, only store log_buf if present.

wcawijngaards committed:
- ixfr-out, routine to construct ixfr file names, and fixup for file counter.

wcawijngaards committed:
- ixfr-out, fix to not log nonexisting files deleted when cleaning up.

wcawijngaards committed:
- ixfr-out, fix to log error on header write failure and whitespace and
  comment text cleanup.

wcawijngaards committed:
- ixfr-out, fix to delete entries if too many before rename could go wrong.

wcawijngaards committed:
- ixfr-out, iterator for ixfr data elements.

wcawijngaards committed:
- ixfr-out, fix delete of remaining files on error.

wcawijngaards committed:
- ixfr-out, fix comment and ixfr_data_prev wraparound.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, fix that rename would not overwrite the wrong files.

wcawijngaards committed:
- ixfr-out, test for ixfr write to file and fixes.

wcawijngaards committed:
- ixfr-out, read for ixfr read from file.

wcawijngaards committed:
- ixfr-out, drop nonlinked IXFRs after an AXFR is received.

wcawijngaards committed:
- ixfr-out, remove unused member variable.

wcawijngaards committed:
- ixfr-out, compression of output.

wcawijngaards committed:
- ixfr-out, fix to not alloc skipped compression pointer insertions.

wcawijngaards committed:
- ixfr-out, limit IXFR packets to the compression bounds of 16k.

wcawijngaards committed:
- ixfr-out, fix variable name.

wcawijngaards committed:
- ixfr-out, fix to copy more data during name compression rdata field steps.

wcawijngaards committed:
- ixfr-out, make get current serial routine public.

wcawijngaards committed:
- ixfr-out, implementation of ixfrcreate, spool to file.

wcawijngaards committed:
- ixfr-out, ixfrcreate iterator implementation.

wcawijngaards committed:
- ixfr-out, ixfrcreate diff implementation.

wcawijngaards committed:
- ixfr-out, move declaration of iterator to header, and faster large domain
  and large RRset comparison.

wcawijngaards committed:
- ixfr-out, ixfr creation, nsd-checkzone -i option to create an ixfr.

wcawijngaards committed:
- ixfr-out, fix nsd-checkzone for ixfr creation. And fix max size 0 is
  unlimited in ixfr size check.

wcawijngaards committed:
- ixfr-out, fix ixfr creation to add domains after old domains end.

wcawijngaards committed:
- ixfr-out, fix ixfr create rename files for more than two files.

wcawijngaards committed:
- ixfr-out, fix ixfr create rename files.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, fix usage for nsd-checkzone -i option to create ixfr contents.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, man page entry for nsd-checkzone -i.

wcawijngaards committed:
- ixfr-out, fix ixfr create to write size in bytes in the ixfr zone file.

wcawijngaards committed:
- ixfr-out, review fixes for cancelled ixfr creation.

wcawijngaards committed:
- ixfr-out, fix to stop ixfr creation when it has already been done.

wcawijngaards committed:
- ixfr-out, remove debug statements.

wcawijngaards committed:
- ixfr-out, extra guard nul on buffer.

wcawijngaards committed:
- ixfr-out, remove timestamps from log, and if server sees ixfr file that
	    has been created, read it up into memory.

wcawijngaards committed:
- ixfr-out, document duplicate check for ixfr creation by nsd-checkzone in
	    the man page for it.

wcawijngaards committed:
- ixfr-out, reformat the nsd-checkzone -i option entry.

wcawijngaards committed:
- ixfr-out, test for nsd-checkzone -i option.

wcawijngaards committed:
- ixfr-out, fix test for nsd-checkzone pre and post.

wcawijngaards committed:
- ixfr-out, fix to delete temporary file in case ixfr creation zone read fails.

wcawijngaards committed:
- ixfr-out, test ixfr creation name, RRs and RRsets with nsd-checkzone.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, create ixfr when zone file is loaded into nsd.

wcawijngaards committed:
- ixfr-out, fix that when no zone is loaded, no difference is created.

wcawijngaards committed:
- ixfr-out, remove -u logname from test to avoid initgroups calls in test.

wcawijngaards committed:
- ixfr-out, test for ixfr create from zone file differences.

wcawijngaards committed:
- ixfr-out, remove debug from test start of ixfrout_create.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, create-ixfr config option.

wcawijngaards committed:
- ixfr-out, nsd-checkzone -n limit the ixfr number.

wcawijngaards committed:
- ixfr-out, an unlink error does not stop ixfrcreate processing.

wcawijngaards committed:
- ixfr-out, make depend for ixfr.c and ixfrcreate.c dependencies.

wcawijngaards committed:
- ixfr-out, implement create-ixfr option. It creates an IXFR when a zonefile
  is read by NSD.

wcawijngaards committed:
- ixfr-out, implement nsd-control -s ixfr size option.

wcawijngaards committed:
- ixfr-out, document the nsd-control -s option.

wcawijngaards committed:
- ixfr-out, move default definitions to ixfr.h.

wcawijngaards committed:
- ixfr-out, make depend for nsd-checkzone dependencies.

wcawijngaards committed:
- ixfr-out, test for create-ixfr option.

wcawijngaards committed:
- ixfr-out, statistics counter for number of IXFR requests, like for AXFR.

wcawijngaards committed:
- ixfr-out, fix assertion for clang analyzer.

wcawijngaards committed:
- ixfr-out, fix assertion again for clang analyzer.

wcawijngaards committed:
- test for nsd-checkzone -i -n (number) option, and fix to make that number.

wcawijngaards committed:
- test for nsd-checkzone -i -s (size) option. And fix error printout text.

wcawijngaards committed:
- ixfr-out, nsd-checkzone -s prints errors, and deletes the files.

wcawijngaards committed:
- ixfr-out, fix comment in ixfr_read_header_data_size.

wcawijngaards committed:
- ixfr-out, fix spelling in struct ixfr_create comment.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wcawijngaards committed:
- ixfr-out, test for rixfr statistic, and fix for axfr in ixfr result.

wcawijngaards committed:
- ixfr-out, improve documentation.

wcawijngaards committed:
- ixfr-out, test that after zonefile from no contents, no ixfr is created.

wcawijngaards committed:
- ixfr-out, add test file.

wcawijngaards committed:
- ixfr-out, test when create-ixfr is not enabled.

k0ekk0ek commented:
Agree on the mechanics. Didn't try to spot programming errors yet since this is still WiP and testcases etc are still coming in. If you want I'll do so at a later stage.

wcawijngaards committed:
- ixfr-out, test for create ixfr in server with ixfr-number option and
  fix deletion of old ixfr data from tree and pick next data element from
  the next tree element if possible.

wcawijngaards committed:
- ixfr-out, test for create ixfr in server with ixfr-size option.

wcawijngaards committed:
- ixfr-out, test for ixfr contents that spans multiple messages on stream.

wcawijngaards committed:
- ixfr-out, test for tsig, and fixes to tsig sign ixfr full zone response

wcawijngaards committed:
- ixfr-out, test for tsig on long ixfr responses. Fix truncation for long
  ixfr responses.

wcawijngaards committed:
- ixfr-out, review fix, do not store offsets beyond the max compression offset.

k0ekk0ek commented on ixfr.c:
```diff
@@ -0,0 +1,2610 @@
+}
+
+/* Count length of next record in data */
+static size_t count_rr_length(uint8_t* data, size_t data_len, size_t current)
```
Maybe `const` for `data`?

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/nsd/pull/209/commits/1454639924f6785a412c0e4577bbf1c8e6c6c14a .

k0ekk0ek commented on ixfr.c:
```diff
@@ -0,0 +1,2612 @@
+/* write an RR into the packet with compression for domain names,
+ * return 0 and resets position if it does not fit in the packet. */
+static int ixfr_write_rr_pkt(struct query* query, struct buffer* packet,
+	struct pktcompression* pcomp, uint8_t* rr, size_t rrlen)
```
Probably best to make `rr` `const`?

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/nsd/pull/209/commits/b458e56e95641d074656b60f37519a4d5d8a00d0 . Thanks for the comments!

k0ekk0ek reviewed (approved):
I've looked quite a bit at the PR. There's some places where the parameters/variables can be `const` (I'm sure I haven't marked all of them). But I haven't found anything that's obviously wrong and the tests pass. My mind is really not a compiler, so I can only trust all the buffer writes at certain offsets are correct. Personally, in places like that I'd combine some portions, e.g. maybe reuse the packet compression code that's currently already in and make it somewhat more generic. Then write a bunch of unit tests to test it more extensively. But I'm fine with it either way.

wcawijngaards committed:
- ixfr-out, review fix, add const in count_rr_length.

wcawijngaards committed:
- ixfr-out, review fix, add const in ixfr_write_rr_pkt.

wcawijngaards committed:
- ixfr-out, test for same serial and fix for same serial in nsd-checkzone ixfr.

wcawijngaards committed:
- ixfr-out, test for same serial from load from file.

wcawijngaards committed:
- ixfr-out, fix to clean up tempzone when reading ixfr file.

wcawijngaards committed:
- ixfr-out, fix to check for child domains before delete when reading ixfr file.

wcawijngaards committed:
- ixfr-out, remove child domains fix that was not necessary, add fix for
  wildcard child closest match that is updated when domains are removed
  when reading ixfr file.

wcawijngaards committed:
- ixfr-out, review fix on wildcard child adjust, when reading ixfr file.

wcawijngaards committed:
- ixfr-out, test for ixfr fetch and ixfr-num.

wcawijngaards committed:
- ixfr-out, test for ixfr fetch and ixfr-size.

wcawijngaards committed:
- ixfr-out, fix the rr owner cleanup during ixfr read.

wcawijngaards committed:
- ixfr-out, fix the rr owner cleanup to not lower usage.

wcawijngaards committed:
- ixfr-out, remove debug from test script.

wcawijngaards committed:
Merge branch 'master' into ixfr-out

wtoorop commented on axfr.c:
```diff
@@ -163,64 +168,78 @@ query_axfr(struct nsd *nsd, struct query *query)
+		DEBUG(DEBUG_XFRD,1, (LOG_INFO, "%s refused, %s",
+			(q->qtype==TYPE_AXFR?"axfr":"ixfr"),
+			acl?"blocked":"no acl matches"));
+		if (!zone_opt) {
+			RCODE_SET(q->packet, RCODE_NOTAUTH);
```
Yes, I can see that we don't need an EDE code here because the rcode is unambiguous, but how is that different from here: https://github.com/NLnetLabs/nsd/blob/ixfr-out/query.c#L1299 . Should we not also set the NOAUTH rcode without the EDE there too? WDYT?
I know it is only moved code b.t.w. So I'm perfectly fine ignoring it. Just wanted to know your opinion.

wcawijngaards replied:
We set the REFUSED rcode in query.c because other implementations also return REFUSED for queries to a not hosted zone. And we want to be compatible.

wtoorop commented on ixfr.h:
```diff
@@ -0,0 +1,270 @@
+	 * has an even newer new_serial of that entry */
+	uint32_t newest_serial;
+	/* the newserial number of the newest ixfr data element */
+	uint32_t newest_newserial;
```
What is it's purpose? It looks like values are assigned to it only, but never actually used. Is that correct?

wcawijngaards replied:
I added it because it allows comparison with the current serial in the nsd domain tree. But that comparison does not happen from that contents. I guess I should remove it, if it is not used.

wcawijngaards replied:
Fixed this in https://github.com/NLnetLabs/nsd/pull/209/commits/d675c33aa921bb319eae12526027f6d97fb2c0c5 . That removes the variable.

wtoorop commented on ixfr.c:
```diff
@@ -0,0 +1,2661 @@
+		if( (lablen&0xc0) )
+			break; /* name should be uncompressed */
+		if(lablen+1 > rrlen)
+			break; /* name should fit */
```
Should these error conditions not return 0 instead of breaking?

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/nsd/pull/209/commits/166394f58efd6f6f41dd7d1f90906a43ac6c3877 . Good to find, even though wrong input is not expected here, the output is then nicer.

But after the commit, I think, what return 0 does, is return that it does not fit. And then it acts as if the RR did not fit. But instead we want to fail the transfer or something.

wcawijngaards replied:
The fix in https://github.com/NLnetLabs/nsd/pull/209/commits/a954ac477c6ad5125033b52c83b62cf9ec615518 skips the RR if there is malformed dname content in it. It returns -1 to signal the malformed data.

wtoorop reviewed (approved):
Thank you Wouter, looks wonderfull!

wcawijngaards committed:
- ixfr-out, review fix in pktcompression_write_dname to return 0 on failure.

wcawijngaards committed:
- ixfr-out, review fix that skips malformed rr for pktcompression_write_dname.

wcawijngaards committed:
- ixfr-out, review fix, remove unused variable with newest serial.
